crates/turborepo-cache/src/cache_archive/create.rs

```diff
@@ -0,0 +1,156 @@
+use std::{
+    backtrace::Backtrace,
+    fs,
+    fs::{File, OpenOptions},
+    io::{BufWriter, Read},
+};
+
+use tar::{EntryType, Header};
+use turbopath::{
+    AbsoluteSystemPathBuf, AnchoredSystemPathBuf, AnchoredUnixPathBuf, RelativeUnixPathBuf,
+};
+
+use crate::CacheError;
+
+enum CacheWriter<'a> {
+    Compressed(tar::Builder<zstd::Encoder<'a, BufWriter<File>>>),
+    Uncompressed(tar::Builder<BufWriter<File>>),
+}
```

Comment:
Unfortunately no, since we can't assign to a variable two different generic implementations, i.e. we can't do:

```
let writer = if compressed {
  tar::Builder::new(zstd::Encoder::new(file)); 
} else {
  tar::Builder::new(file)
};
```

In fact, I believe I threw in the towel in a later PR and just made this a `Box<dyn Write>` which is a lot better tbh, at the cost of some small performance impact.

Reply:
I saw that this one already had `Box<dyn Read>` and I'm generally in favor of this. 👍